### DISCUSSION TITLE
needs_string_links for different cases, i.e. generating different links depending on content

### DISCUSSION THREAD

**Discussion by hcc23** [useblocks/sphinx-needs#911] [Ideas]
Hi,

say I have two needs of the same type, such as this:

```
.. need:: Need 1
    :id: 1
    :owner: john-doe_user

.. need:: Need 2
    :id: 2
    :owner: jane.doe@mail.com
```

Note that the format of the option entry for `owner` differs, the one in Need 1 is meant to represent a github user identification, the one in Need 2 is a plain email address.

According to [1](https://sphinx-needs.readthedocs.io/en/latest/configuration.html#needs-string-links) there exists the option to render the owner option as a link when processed by sphinx.  I want to allow both formats, but still want clickable links, hence I am hoping to write clever `needs_link_strings` to capture these cases*:

```
needs_string_links = {
    'render_github_user': {
        'regex': r"^(?P<value>.+_user)$",  # hypotheses: all github users end in _user
        'link_url': "https://github.com/orgs/MyOrg/teams/" + "{{value}}",
        'link_name': "{{value}}",
        'options': ['owner']
    },
    'render_email_address': {
        'regex': r"^(?P<value>.+@.+)$",  # poor man's email-regex
        'link_url' : "mailto:" + "{{value}}",
        'link_name': "{{value}}"
        'options': ['owner']
    },
}
```

In my experience this doesn't seem to work as it _appears_ (I wasn't able to find conclusive evidence on this) that only the first entry is pulled, i.e. the first dict pertaining to the `:owner:` option is consumed and, if the regex doesn't match, nothing happens. Which is ok -- but any subsequent dicts also corresponding to the `:owner:` element don't seem to be taken into account.

Is there a chance that this "mulitple entry" approach could be added or "hacked around" somehow?

--
*: yes, I know that the regexes aren't the nicest ones around, but please bear with me :)

PS: I have also posted this - formed as a question - at StackOverflow: https://stackoverflow.com/q/76047208/1984137

**Comment by HughG** [comment]
I will probably also want to do this, but don't have time to play with it right now.  Could you put an artificial prefix on each kind of format, and pull that out in the regex? Having a distinct prefix might avoid the wrong regex greedily matching.  E.g.,

```python
needs_string_links = {
    'render_github_user': {
        'regex': r"^github:(?P<value>.+)$",
        'link_url': "https://github.com/orgs/MyOrg/teams/" + "{{value}}",
        'link_name': "{{value}}",
        'options': ['owner']
    },
    'render_email_address': {
        'regex': r"^mailto:(?P<value>.+)$",
        'link_url' : "mailto:" + "{{value}}",
        'link_name': "{{value}}"
        'options': ['owner']
    },
}
```

Then in your needs, have
```rst
.. need:: Need 1
    :id: 1
    :owner: github:john-doe_user

.. need:: Need 2
    :id: 2
    :owner: mailto:jane.doe@mail.com
```
